plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java

```diff
@@ -61,18 +64,40 @@
     // https://cloud.google.com/storage/docs/json_api/v1/how-tos/resumable-upload
     private static final int LARGE_BLOB_THRESHOLD_BYTE_SIZE = 5 * 1024 * 1024;
 
-    private final Storage storage;
-    private final String bucket;
+    private final String bucketName;
+    private final String clientName;
+    private final GoogleCloudStorageService storageService;
 
-    GoogleCloudStorageBlobStore(Settings settings, String bucket, Storage storage) {
+    GoogleCloudStorageBlobStore(Settings settings, String bucketName, String clientName, GoogleCloudStorageService storageService) {
         super(settings);
-        this.bucket = bucket;
-        this.storage = storage;
-        if (doesBucketExist(bucket) == false) {
-            throw new BlobStoreException("Bucket [" + bucket + "] does not exist");
+        this.bucketName = bucketName;
+        this.clientName = clientName;
+        this.storageService = storageService;
+        if (doesBucketExist(bucketName) == false) {
+            throw new BlobStoreException("Bucket [" + bucketName + "] does not exist");
         }
     }
 
+    private <R, E extends IOException> R storageAccess(CheckedFunction<Storage, R, E> storageFunction) throws IOException {
```

Comment:
@tlrx I kinda think the extra noise is worth the trouble.

For one thing, trying to locally save the reference (that you get as an argument to the lambda) is an obvious code smell. We convey that you only borrow the reference as an argument to the lambda, you don't get to store it locally. In the [S3 repo plugin the simple approach was vetoed](https://github.com/elastic/elasticsearch/blob/acbefac5584df3e1f8d1a1cd0a6388e21d1a0230/plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java#L72), but there the client was closeable, so the IDE hinted a resource leak.
Secondly, using the noisy approach we are cutting the access control stack **after** the client has been built. That is because, if the client is not cached it has to be created, which generally speaking, does not require elevated privileges. In the simple approach the elevated privilege encompass the client creation too. This is however only a theoretical benefit, since currently we sadly [elevate privileges when creating the client](https://github.com/elastic/elasticsearch/blob/acbefac5584df3e1f8d1a1cd0a6388e21d1a0230/plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRepository.java#L89) (to allow for the google default application credentials).

These were my thoughts, but sometimes I tend to complicate things.
Do you wish to sudo?

Reply:
I don't want to sudo, I just want to express my concerns about this. GoogleCloudStorageService provides Storage clients to the plugin but they have to be used in a specific way and not be stored locally, that's why we are adding storageAccess() methods like this in the BlobStore implementation. I'm tempted to think that we should not expose Storage clients at all and just have GoogleCloudStorageService (or something else) that borrows a temporary client to execute the needed operation and just returns the result. So that we just keep Storage internal. WDYT?

Reply:
Ok. So we could pack the `GoogleCloudStorageService` together with the `clientName` in another entity, and this would be responsible for handling the operations without exposing the `Storage`. Yet, I dislike this because I feel like this encapsulation of the client is already done at the `BlobStore` level and away from the `BlobContainer`; `BlobContainer` implements blob operations, abstracting away the client. This approach, if I got it right, will nest the client even further under and possibly stripping `BlobStore` or `BlobContainer` of their duties. In this matter, I think [`AzureBlobContainer`](https://github.com/elastic/elasticsearch/blob/master/plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobContainer.java) is illustrative; `AzureBlobStore` and `AzureBlobContainer` could be collapsed into a single class. This is in part because the azure client is relatively high level and the encapsulation we are talking about is handled internally.

Writing this down, I am gravitating towards your suggestion. There still has to be someone that knows how to use the client (not to cache it locally) and adding another layer (entity or method receiving lambda) is without justification, `BlobStore` should handle it.

Unless you have other thoughts I will go ahead and have a simple private method inside the `BlobStore` that returns the `Storage` instance. **Code inside the `Blobstore` is responsible for not caching the instance.**

Reply:
There is history around the azure plugin so I won't take this as an example. 

> Unless you have other thoughts I will go ahead and have a simple private method inside the BlobStore that returns the Storage instance. Code inside the Blobstore is responsible for not caching the instance.

:+1: 

Let's do that and not block this PR. This is something we can still revisit later on.